### PR TITLE
caffe: update to 1de4ceb (20170817)

### PR DESCRIPTION
**Pull request by GerardSoleCa**
###### Description
Port for caffe was old (2015), so this commit upgrades its source to
current master stable. It updates required patches so library will
install successfuly.
Dynamic libraries are amended, so they will point to the correct
libcaffe. Current Port, has an issue with python and caffe, that
requires manual recolocation of affected library.
Also, tests are added for both library and python module.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on

macOS 10.12.5 16F73
Xcode 8.3.3 8E3004b

###### Verification 
Have you
- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?

- [x] checked your Portfile with `port lint`?
- [x] tried existing tests with `sudo port test`?
- [x] tried a full install with `sudo port -vst install`?
- [x] tested basic functionality of all binary files?

### CODE REVIEW

File: math/caffe/Portfile [hunk 1]
```diff
@@ -3,9 +3,10 @@
 PortSystem          1.0
 PortGroup           github 1.0
 
-github.setup        BVLC caffe 6d92d8fcfe0eea9495ffbc326256ec5b70c3eed1
-version             20150717
-revision            6
+name                caffe
+github.setup        BVLC caffe 1de4cebfb81d50267d0d8c2595372b14e1408248
+version             20170823
+revision            7
```
Comment (l2dy) [inline math/caffe/Portfile]: `revision` should be cleared when the port is updated to a newer version.

Comment (GerardSoleCa) [reply math/caffe/Portfile]: Removing it, missed that part from the documentation, Thanks! 😄

**Comment by l2dy** [comment]
The version in the Portfile (20170823) and the commit message (20170826) doesn't match.

**Comment by GerardSoleCa** [comment]
Portfile, commit message and PR updated. 
Using now the day of caffe commit version, so now all should be pointing to the day of the commit in caffe repository.

File: math/caffe/files/patch-cudnn.diff [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
++++ Makefile.config	2017-08-24 10:43:54.000000000 +0900
 @@ -2,7 +2,7 @@
  # Contributions simplifying and improving our build system are welcome!
- 
+
  # cuDNN acceleration switch (uncomment to build with cuDNN).
 -# USE_CUDNN := 1
-+USE_CUDNN := 1
- 
++ USE_CUDNN := 1
```
Comment (l2dy) [inline math/caffe/files/patch-cudnn.diff]: Why an extra space?

File: math/caffe/files/patch-Makefile.config.diff [hunk 3]
```diff
@@ -1,7 +1,16 @@
...
-@@ -48,8 +48,9 @@
- 
+--- Makefile.config.orig	2017-08-24 10:43:43.000000000 +0900
++++ Makefile.config	2017-08-24 10:43:54.000000000 +0900
+@@ -18,7 +18,7 @@
+ # ALLOW_LMDB_NOLOCK := 1
+
+ # Uncomment if you're using OpenCV 3
+-# OPENCV_VERSION := 3
++ OPENCV_VERSION := 3
```
Comment (l2dy) [inline math/caffe/files/patch-Makefile.config.diff]: Same.

File: math/caffe/files/patch-cpu-only.diff [hunk 4]
```diff
@@ -1,37 +1,45 @@
...
  # USE_CUDNN := 1
- 
+
  # CPU-only switch (uncomment to build without GPU support).
 -# CPU_ONLY := 1
-+CPU_ONLY := 1
- 
- # To customize your choice of compiler, uncomment and set the following.
- # N.B. the default for Linux is g++ and the default for OSX is clang++
++ CPU_ONLY := 1
```
Comment (l2dy) [inline math/caffe/files/patch-cpu-only.diff]: Same.

File: math/caffe/files/patch-cudnn.diff (hunk 2 above)
Comment (GerardSoleCa) [reply math/caffe/files/patch-cudnn.diff]: My fault, when I removed the commend hash by hand on the original source code, but missed the space between the hash and the variable.
I've summited the removal of that spaces.

File: math/caffe/files/patch-cudnn.diff [hunk 5]
```diff
@@ -1,11 +1,11 @@
...
+--- Makefile.config.orig	2017-08-24 10:43:43.000000000 +0900
++++ Makefile.config	2017-08-24 10:43:54.000000000 +0900
 @@ -2,7 +2,7 @@
  # Contributions simplifying and improving our build system are welcome!
- 
+
  # cuDNN acceleration switch (uncomment to build with cuDNN).
 -# USE_CUDNN := 1
 +USE_CUDNN := 1
-
```
Comment (l2dy) [inline math/caffe/files/patch-cudnn.diff]: This space came from `diff -u`, why did you remove it?

File: math/caffe/Portfile [hunk 6]
```diff
@@ -117,18 +125,18 @@ variant python27 description {Install Python 2.7 interface} {
                     port:py27-protobuf \
                     port:py27-gflags \
                     port:py27-leveldb \
-                    port:py27-dateutil
+                    port:py27-dateutil \
+                    port:py27-pydot
 
     build.target-append  pycaffe
+    test.target-append  pytest
```
Comment (l2dy) [inline math/caffe/Portfile]: Little nitpick, you can add one more space before `pytest` to align it with the line above.

File: math/caffe/files/patch-cudnn.diff (hunk 5 above)
Comment (GerardSoleCa) [reply math/caffe/files/patch-cudnn.diff]: Before opening the PR, when I was upgrading the whole thing, I generated all the patches again with 
```diff -Naur original_file new_file > patch.txt```
So the removal of that spaces are from diff -Naur.

I've been looking to caffe source, and it seems that that lines doesn't contain any space (https://github.com/BVLC/caffe/blob/master/Makefile.config.example#L3 and https://github.com/BVLC/caffe/blob/master/Makefile.config.example#L6). So in fact, I think that it's the normal behaviour of diff, to just keep empty line.

And I think that the patch on master has does spaced which should not be. I've also checked against the sources that caffe port is pointing at, and neither on that lines there is no space: https://github.com/BVLC/caffe/blob/6d92d8fcfe0eea9495ffbc326256ec5b70c3eed1/Makefile.config.example#L6

So, I understand that no space should be on the diff file.

File: math/caffe/Portfile (hunk 6 above)
Comment (GerardSoleCa) [reply math/caffe/Portfile]: 👍

File: math/caffe/files/patch-cudnn.diff (hunk 5 above)
Comment (l2dy) [reply math/caffe/files/patch-cudnn.diff]: > The lines common to both files begin with a space character in unified format.

I tried `diff -Naur` but these spaces indeed exist, did you use an editor that automatically removes trailing whitespaces?

Comment (GerardSoleCa) [reply math/caffe/files/patch-cudnn.diff]: Uhm, damn.., seems that Atom removes trailing whitespaces.
Added all of them with vi, so now should follow diff spec.
Thanks for the link!
